### PR TITLE
Pull-up with recent master

### PR DESCRIPTION
**Pull request by MarcelHB**
Now I noticed the full degree of disruption by my recent changes.

For the first step, I merged the current HEAD and hammered things around to make them work and compile green. But since `GemRB_Window_SetPicture` already disappeared and my IWD2 night-menu screen changes made use of that, the main menu is already a little broken. I guess I will have to restore that method. :)

### CODE REVIEW

**Comment by bradallred** [comment]
oh wow! Just about to hit the sack here. I'll have to digest this over lunch tomorrow. Looks like a lot of good work, so thank you :)

Edit: is this up-to-date with the main master now?

**Comment by MarcelHB** [comment]
It's as up-to-date as can be. I was not really sure how to fully solve some conflicting things around SDL but since your are in WIP there as well, I guess this will need further work anyway.

**Comment by bradallred** [comment]
ah, ok. I imagine you are referring to SDL2 conflicts; I don't recall any SDL changes outside that. I believe all those conflicts can simply be discarded (most changes there were complete hacks or addressed elsewhere now).

Were you able to build with SDL 1.2 for now, or should I prioritize getting the SDL 2 driver reassembled for you?

**Comment by bradallred** [comment]
I believe that `GemRB_Window_SetPicture` is simply replaced with `GemRB_View_SetBackground`.

If you require additional customization for the `Window` version of that; doing that on the Python side would be better then having a GUIScript method for it. There are several examples of this in `GUIClasses.py`

**Comment by MarcelHB** [comment]
Ok, good to know.

SDL 1.2 works well, as long as we talk about my Linux desktop VM.

File: gemrb/core/GUI/EventMgr.h [hunk 1]
```diff
@@ -74,6 +74,8 @@ class Window;
 #define GEM_MB_SCRLUP           8
 #define GEM_MB_SCRLDOWN         16
 
+#define GEM_MB_ONGOING_ACTION   33
```
Comment (bradallred) [inline gemrb/core/GUI/EventMgr.h]: This looks unused? Also the new Event struct has a repeat count field and possibly other useful attributes that could/should be used instead I think.

If it would be helpful, we can alter the `OnKeyX` and `OnMouseX` methods to directly take the event object so all that stuff is available there. I've already (locally) changed the key handlers to take the `KeyboardEvent` directly, but it would be no big thing to use the raw event instead; thats what most GUI libraries seem to do anyway.

File: gemrb/plugins/GUIScript/GUIScript.cpp [hunk 2]
```diff
@@ -1638,6 +1638,48 @@ static PyObject* GemRB_Control_SetText(PyObject* self, PyObject* args)
 	Py_RETURN_NONE;
 }
 
+PyDoc_STRVAR( GemRB_Window_SetKeyPressEvent__doc,
```
Comment (bradallred) [inline gemrb/plugins/GUIScript/GUIScript.cpp:1641]: this is at the very least going to need a rewrite since we don't have a "window index" anymore. nor does `core` have APIs for getting at the windows.

Can't this be handled by `Window::RegisterHotKeyCallback`? We have a Python variant for `EventCallback`

If you are just trying to use this with buttons then `button.SetHotkey` should do what is needed already.

File: gemrb/plugins/GUIScript/PythonHelpers.cpp [hunk 3]
```diff
@@ -111,3 +75,62 @@ void PythonControlCallback::operator() (Control* ctrl) const
 	Py_DECREF(co_argcount);
 	CallPython(Function, args);
 }
+
+template <>
+bool PythonObjectCallback<WindowKeyPress>::operator() (WindowKeyPress *wkp) {
```
Comment (bradallred) [inline gemrb/plugins/GUIScript/PythonHelpers.cpp]: This too ought to already be handled by the existing python callbacks. The python representation of the window (`GWindow`) would be passed as the parameter to the python function.

**Comment by bradallred** [comment]
Most of the (C++) Python changes look good (left a note about one), but I'm mostly trusting that it works :p 
this being a WIP branch I'm not concerned about breaking enough to thoroughly vet it.

**Comment by MarcelHB** [comment]
Ok, I hope I can go through your comments tomorrow.

**Comment by bradallred** [comment]
So looking between this and the pull requests from the main repo, it looks like all of the Python and window event changes are only there to support activating PCs with the number keys. The equivalent on this branch would normally simply be to call `SetHotkey` on the buttons in the same loop that is doing the portrait window setup...

Pretty sure that won't work as is, however, since it probably just set the hotkey for when the button's window has focus. What we _could_ do here is have `GameControl` setup a callback with `EventMgr:: RegisterEventMonitor`, but that has an unwanted side-effect of changing selected PC even when just trying to type a number in an input field...

Perhaps the best approach would be to funnel _unhandled_ key events to the `WindowManager:: gameWin` which would end up being delivered to `GameControl` and that would work, but then GameControl could receive events while a modal window is displaying... which we actually would want for the PC selection, but nothing else (that I know of). So the callback would have to check `WindowManager:: IsPresentingModalWindow` and bail for non-numeric keys.

That might have a nice sideeffect of being able to have all the hot key logic for both the ActionBar and PortraitWindow in a single concise location (inside `GameControl`) rather then sprinkled about in the Python scripts (likely duplicated all over the place) with funky one-off GUIScript APIs

What do you think? I'll play around tonight and see if I can get something pushed up to discuss.

**Comment by bradallred** [comment]
so... it turns out that apparently I already thought of this and it actually works right now (I had to push a one line change to ensure the `GameControl` is the focus of its window).

The problem is that the window's python "update script" needs to get triggered...

It would probably be cleanest to give `View` a refresh event handler that we could call everytime the `MarkDirty` is called (probably have it check the existing value of the dirty flag to avoid extra calls). This would also be much less manual then we currently do too... I'm pretty sure there are several places where we assign an event call back just so we can call the update function. Might fix a few bugs too (like char hp not refreshing while looking in the inventory etc)

**Comment by bradallred** [comment]
nevermind 😀 

turns out the failure to refresh was a missing `SelectPCSingle` call. It was actually your pull request on the main repo that made me realize that! so thank you.

tl;dr
Using the number buttons to make PC selections seems to work fine here now.

File: gemrb/plugins/GUIScript/GUIScript.cpp (hunk 2 above)
Comment (MarcelHB) [inline gemrb/plugins/GUIScript/GUIScript.cpp:1641]: I've updated the doc, but it is already rewritten internally. If not needed anymore, then we can also purge it.

**Comment by MarcelHB** [comment]
Ok, I have taken care of the minor things right now.

I have already spotted some weird displacements of button windows in BG, but I suggest to take care of that in the next steps.

**Comment by bradallred** [comment]
I set up Travis today to help fix the compile warnings. I probably should find the time to fix the 100 or so precision loss and dead code warnings (a few others too) in the main repo so that I can treat warnings as errors again on my local machine.

I haven't noticed any "weird displacements", but perhaps I simply need to run with a different resolution (I mainly run 640x480 so I can run and debug easily on my laptop). I will take a look when I get home.

I think the final question to answer before I merge this is: what is the purpose of `Window_SetKeyPressEvent` and `WindowKeyPressHandler`? They seem superfluous in this branch...

**Comment by MarcelHB** [comment]
Going to test this tonight, otherwise I will remove it.

**Comment by bradallred** [comment]
Can you enlighten me as to what 4758aa3 was specifically supposed to fix? there isn't much to go on in the commit message.

Apparently that conflicted with changes I had made and now with new changes I just pushed up. Was this tested? I'm concerned your merge strategy was to select your changes then fixup the compile errors and I'm uncertain if anything important of mine got dropped or misplaced 😛 

Like I said, not too concerned with breaking anything on this branch, I'd just rather know what the intent was and if you tested it (not that everything was working 100% there anyway).

**Comment by MarcelHB** [comment]
Sorry, was a little out of time.

I just introduced a new event-bit there to have some distinction between the physical `OnMouseUp` events and that fake one to make pressed keys work as if e.g. buttons were pressed again. This was necessary to make the scrollbar work in the CG of IWD2 after enabling repeating keys as it lost focus immediately, rendering dragging impossible.

IWD2 was back online, but I didn't test that specific fix in your branch yet.

So the answer is: I just rewrote the bit-pattern checks to remain compatible.

**Comment by bradallred** [comment]
I have/will be removing all of our phony events in this branch so that wont be necessary (and likely doesn't work here now). It never was a good way of doing things, it was just an easy shortcut we used because of existing poor design.

The repeat event stuff will be moved directly into the `Control` class eventually. Really just need to refactor some of the event handlers into a common "action" handler. Once that happens a mouse down event can start a timer that repeatedly fires the controls action callback and a mouse up will cancel that timer. Much less hacky and will lend itself to cleaner more modular code rather then our tangled event handling methods we have now.

The focus problems are a separate issue, but those too should be addressed here in a different way then we have historically, if it isn't already.

The entire scrollbar -> TA or Window relationship was scrapped and that behavior is being rewritten in the `ScrollView` class. `Scrollbar` only needs to be a dumb class that reflects the state of something else (usually a `ScrollView`) and fires a callback when it changes its value so the lister can update its state rather then the scrollbar actually knowing about and doing that directly.

I've been pushing up bits and pieces of that here and there, but I'm still working on some details that aren't quite ready yet.

**Comment by MarcelHB** [comment]
Ah, I remember ... IWD2 load screen gets stuck by some Python exceptions. Gonna take a look into them.

**Comment by MarcelHB** [comment]
By the way, this is my BG on this branch right now.
screenshot_bg

**Comment by MarcelHB** [comment]
Same on your `subviews` branch, ScollBars (IWD2) work nowhere as well so far.

In the scripts, lots of things seem to be broken around the dropped `LoadWindowPack` and my attempts to play with `LoadWindow` instead only caused more problems deeper down the stack, such as `GetControl`s failing to get anything.

And I don't get the Travis error. Is it some static analysis setting? Here, everything looks good, for evident reasons.

**Comment by bradallred** [comment]
Right, I thought I mentioned in my email that _no other games_ besides BG2 have had their GUIScripts updated (and even then BG2 still has some left to do). That is why you see what you see in BG and why IWD2 is throwing exceptions. Those windows clustered in the center just need to have their position set like their counterparts in `GUIScripts/bg2/MessageWindow.py`; look for `WINDOW_LEFT` and such. It looks like other problems are occurring as well tho.

Yes, `LoadWindowPack` is gone and replaced with `LoadWindow`. If you compare the BG2 counterpart scripts you should see that you can delete _a lot_ of code in the `OpenXWindow` functions and replace it with a few lines. Look for `GUICommonWindows.OpenTopWindow`.

The window management on the Python side still needs some work. I want to basically make a window manager that we can just push windows onto and when we close them it should automatically refresh the new top window automatically. right now all of the subwindows are completely done manually with lots of global variables tracking the windows; real messy stuff. We should be able to rid ourselves of that easily now. This is probably my next area of focus along with the `ScrollView`

Re scrollbars: they have been disconnected right now because I am working on the `ScrollView` class. There is no longer an easy way to reconnect them since I have now removed the `Scrollbar` member from `View`. It will be a while before they work again :(

There should be plenty that you can do for IWD2 or whatever other game without scrollbars I would think :) Really just getting it to the point where the game loads up correctly would be a godsend and 95+% of the work unique to IWD/IWD2/BG; PST may need a bit more love, I don't remember the status of its little floating window menu.

The Travis Error is my fault: `error: assigning to 'unsigned short' from incompatible type 'const GemRB::KeyboardEvent' wkp.key = key;`

The Travis wont work on my branch until this is merged in because I'm still missing some Travis stuff from main master.

This pull request is getting too long as it is. I'm just going to merge it in and revert the stuff that looked unneeded for now. We can pull those bits back in if we need them later on.

Thanks again for all your help with this 😃

**Comment by MarcelHB** [comment]
:+1:

Will go through IWD2/BG over time and find things to fix that will hopefully not interfere with your work.
